### PR TITLE
Cache ratings for Section objects

### PR DESCRIPTION
**Pull request by ZiqiXiong**
After my unsuccessful attempt (https://github.com/aspc/mainsite/pull/117) to improve Course Planner's performance by GenericSection, I discarded GenericSection and tried the simple approach of caching ratings as FloatFields in Section models, which cut down the running time by almost half on my localhost.

### CODE REVIEW

**Comment by mattdahl** [comment]
This works but I just don't think it's very clean. We're basically replicating the same data across multiple "Section" objects -- for every course that has multiple sections (which is literally every course unless this semester is the first time that the course is being offered and it only has one section), we are storing the exact same rating data on the corresponding "Section" object. E.g. sections 01, 02, and 03 of Course A will all have the same rating data (and will all need to be updated concurrently when a new review is published that affects them). Again, this works, but it seems a bit like using a bazooka to kill a fly.

The only logical problem that is see is here: https://github.com/aspc/mainsite/pull/118/files#diff-7f5492035f4336fa7339367513d81501R421. It's not enough to merely consider only sections offered this semester as "affected." What if, for example, no sections of the course are being offered this semester, but a user wants to add a review since he took that course last semester? The rating data on that review wouldn't propagate to the necessary "Section" objects properly. I think the "affected sections" would have to be every single section associated with the course. Moreover, the limitation of our current code is that we can't deterministically say which "Section" object will be the one to actually display its rating data (https://github.com/aspc/mainsite/blob/master/aspc/courses/views.py#L317). So I think to be safe we really would have to replicate the rating data across ALL the "Section" objects, not just the ones in the current term.

**Comment by ZiqiXiong** [comment]
The current implementation always gives the most recent section in SectionDetails view, so affected section can only be the most recent one. @mattdahl

**Comment by mattdahl** [comment]
It's supposed to be the most recent section, but how guaranteed is that? If, for instance, there are two sections "tied" for being most recent (insofar as they both took place during the most recent semester, which is the most detail that we have about timing), will sorting by term and then taking the first one always return the same object? Or will sometimes one return, and sometimes the other, since both qualify as "most recent" interchangeably? I'm just not sure.

**Comment by ZiqiXiong** [comment]
@mattdahl  I think the way Django breaks tie is deterministic. Might do more research about that.

**Comment by ZiqiXiong** [comment]
I don't think there is any possible randomness in the most recent section any more, given that no two sections in one term shares the same code slug. Can you think of any other way this may go wrong? @mattdahl

File: aspc/courses/models.py [hunk 1]
```diff
@@ -67,6 +67,17 @@ def slug(self):
     def get_miscellaneous_ratings(self):
         return [self.useful_rating or 0, self.engagement_rating or 0, self.difficulty_rating or 0, self.competency_rating or 0, self.lecturing_rating or 0, self.enthusiasm_rating or 0, self.approachable_rating or 0]
 
+    def update_ratings(self):
+        instructor_related_reviews = CourseReview.objects.filter(instructor = self)
+        self.rating = instructor_related_reviews.aggregate(Avg("overall_rating"))["overall_rating__avg"]
+        self.useful_rating = instructor_related_reviews.aggregate(Avg("useful_rating"))["useful_rating__avg"]
```
Comment (KentShikama) [inline aspc/courses/models.py]: @ZiqiXiong I'm wonder if all these aggregations can't be done at once?

Pseudo-SQL:

SELECT Avg("overall_rating"), Avg("useful_rating"), etc FROM reviews whatever WHERE reviews.instructor_id == self.id

Comment (KentShikama) [inline aspc/courses/models.py]: See https://docs.djangoproject.com/en/1.10/topics/db/sql/

File: aspc/courses/models.py [hunk 2]
```diff
@@ -382,28 +418,15 @@ def __unicode__(self):
     def get_url_to_section_page(self):
         return '/courses/browse/instructor/{0}/course/{1}/'.format(self.instructor.id, self.course.code_slug)
 
+    def _get_affected_sections(self):
+        current_term = Term.objects.all()[0]
```
Comment (KentShikama) [inline aspc/courses/models.py]: @ZiqiXiong Term.objects.first() or is it not the first?

File: aspc/courses/models.py [hunk 3]
```diff
@@ -382,28 +418,15 @@ def __unicode__(self):
...
-        self.course.engagement_rating = course_related_reviews.aggregate(Avg("engagement_rating"))["engagement_rating__avg"]
-        self.course.difficulty_rating = course_related_reviews.aggregate(Avg("difficulty_rating"))["difficulty_rating__avg"]
-        self.course.competency_rating = course_related_reviews.aggregate(Avg("competency_rating"))["competency_rating__avg"]
-        self.course.lecturing_rating = course_related_reviews.aggregate(Avg("lecturing_rating"))["lecturing_rating__avg"]
-        self.course.enthusiasm_rating = course_related_reviews.aggregate(Avg("enthusiasm_rating"))["enthusiasm_rating__avg"]
-        self.course.approachable_rating = course_related_reviews.aggregate(Avg("approachable_rating"))["approachable_rating__avg"]
-        self.course.save()
+        self.instructor.update_ratings()
+        self.course.update_ratings()
+        section = Section.objects.filter(course=self.course, instructors=self.instructor).order_by('term','code_slug')[0]
```
Comment (KentShikama) [inline aspc/courses/models.py]: Same here .first() or not?

File: aspc/courses/management/commands/import_courses.py [hunk 4]
```diff
@@ -175,6 +175,7 @@ def _create_or_update_section(self, section_data, course_object):
 
 		# Save the fully-formed Section
 		section_object.save()
```
Comment (KentShikama) [inline aspc/courses/management/commands/import_courses.py]: @ZiqiXiong Is it possible to remove this .save() since you're saving again below?

Comment (ZiqiXiong) [inline aspc/courses/management/commands/import_courses.py]: No, we need to save the section object first to access its instructors in update_ratings.

**Comment by KentShikama** [comment]
LGTM 👍

**Pull request merged**